native/cocos/bindings/sebind/intl/common.h

```diff
@@ -0,0 +1,1003 @@
+
+            return self->constructWithTuple(remapArgs, n);
+        }
+    };
+    template <>
+    struct ConstructFn<false> {
+        template <typename Self, typename... ARGS2, size_t... indexes>
+        static auto invoke(Self *self, se::Object * /*thisObj*/, std::tuple<ARGS2...> &args, std::index_sequence<indexes...> n) {
+            return self->constructWithTupleValue(args, n);
+        }
+    };
+
+    bool construct(se::State &state) override {
+        using type_mapping = TypeMapping<TypeList<ARGS...>>;
+        using args_holder_type = typename MapTypeListToTuple<typename type_mapping::input_types>::tuple;
+        se::PrivateObjectBase *self{nullptr};
+        se::Object *thisObj = state.thisObject();
+        args_holder_type args{};
+        const auto &jsArgs = state.args();
+        convert_js_args_to_tuple(jsArgs, args, thisObj, std::make_index_sequence<type_mapping::NEW_ARGN>());
+        self = ConstructFn<type_mapping::NEED_REMAP>::invoke(this, thisObj, args, std::make_index_sequence<type_mapping::FULL_ARGN>{});
+        state.thisObject()->setPrivateObject(self);
+        return true;
+    }
+
+    template <typename... ARGS_HT, size_t... indexes>
+    se::PrivateObjectBase *constructWithTuple(std::tuple<ARGS_HT...> &args, std::index_sequence<indexes...> /*unused*/) {
+        return JSB_MAKE_PRIVATE_OBJECT(T, std::get<indexes>(args)...);
+    }
+
+    template <typename... ARGS_HT, size_t... indexes>
+    se::PrivateObjectBase *constructWithTupleValue(std::tuple<ARGS_HT...> &args, std::index_sequence<indexes...> /*unused*/) {
+        return JSB_MAKE_PRIVATE_OBJECT(T, std::get<indexes>(args).value()...);
+    }
+};
+
+template <typename T, typename... ARGS>
+struct Constructor<T *(*)(ARGS...)> : ConstructorBase {
+    using type = T *(*)(ARGS...);
+    type func;
+    bool construct(se::State &state) override {
+        if ((sizeof...(ARGS)) != state.args().size()) {
+            return false;
+        }
+        std::tuple<HolderType<ARGS, std::is_reference<ARGS>::value>...> args{};
+        const auto &jsArgs = state.args();
+        se::Object *thisObj = state.thisObject();
+        convert_js_args_to_tuple(jsArgs, args, thisObj, std::make_index_sequence<sizeof...(ARGS)>());
+        T *ptr = constructWithTuple(args, std::make_index_sequence<sizeof...(ARGS)>());
+        state.thisObject()->setPrivateData(ptr);
+        return true;
+    }
+
+    template <typename... ARGS_HT, size_t... indexes>
+    T *constructWithTuple(std::tuple<ARGS_HT...> &args, std::index_sequence<indexes...> /*unused*/) {
+        return (*func)(std::get<indexes>(args).value()...);
+    }
+};
+
+template <typename T>
+struct Finalizer : FinalizerBase {
+    using type = void (*)(T *);
+    using arg_type = T;
+    type func;
+    void finalize(void *ptr) override {
+        (*func)(reinterpret_cast<T *>(ptr));
+    }
+};
+
+template <typename T, typename R, typename... ARGS>
+struct InstanceMethod<R (T::*)(ARGS...)> : InstanceMethodBase {
```

Comment:
Is it better to use public inheritance?
```c++
template <typename T, typename R, typename... ARGS>
struct InstanceMethod<R (T::*)(ARGS...)> : public InstanceMethodBase {
```

Then if we get a `InstanceMethodBase` pointer, we could invoke its virtual functions.

Reply:
struct inheritance is public by default, the `public` keyword is omited here.